File: leetcode3/이진희/713_Subarray Product Less Than K.java
/*

1. 아이디어 : 주어진 배열에서 모든 원소의 곱이 k를 넘지 않는 연속된 부분배열의 개수를 구한다.
              1. 완전탐색으로 배열의 오른쪽 숫자를 순차적으로 곱함
              2. 곱이 k개 이상이면 왼쪽 숫자를 뺌
              3. 최종적으로 right-left+1를 하여 현재 만들수있는 연속 부분배열을 찾아서 더함

2. 시간복잡도 : O(N)

3. 자료구조/알고리즘 : 슬라이딩 윈도우

 */

class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        // 곱이 k보다 작은 nums의 연속 부분집합
        // 전부 양수 

        if(k <= 1) return 0;

        int ans = 0;
        int l = 0;
        int sum = 1;
        
        for(int r=0; r<nums.length; r++) {
            sum*=nums[r];

            while(sum >= k) {
                sum/=nums[l];
                l++;
            }

            int cnt = r-l+1;
            ans+=cnt;
        }

        return ans;
    }
}

File: leetcode3/최원준/713. Subarray Product Less Than K.py
#

'''
1. 아이디어 :
투 포인터.
right는 하나씩 옮기면서 curr 값에 곱한다.
curr 값이 k가 넘어가면 left를 옮기면서 curr값을 나눈다.
ans += right - left + 1. (오른쪽 포인터의 숫자를 기준으로 갯수를 더함)

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :
투포인터

'''

from collections import defaultdict
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        if k <= 1:
            return 0
        """
        [10, 5, 2, 6, 12, 9, 4, 3, 1, 100, 3] 100

        """
        ans = 0
        n = len(nums)

        left = 0
        right = 0
        curr = 1

        while right < n:
            curr *= nums[right]

            while curr >= k:
                curr /= nums[left]
                left+=1
                
            ans += right - left + 1
            right+=1
        return ans


